### ISSUE TITLE
Backup and restore support

### ISSUE THREAD

**Issue by nadavkav** [PoetOS/moodle-local_metadata#4]
Any ideas about this? as we are about to start implementing it.

**Comment by mchurchward** [comment]
Nope. I haven't come up with any way to support this except through core changes.

**Comment by nadavkav** [comment]
Should we open an MDL in the tracker for those core changes? are you happy with them?

**Comment by emdalton** [comment]
Good point, we really need backup/restore support. Maybe this can be part of the impetus for making this part of Core (or implementing hooks... ;) )

**Comment by mchurchward** [comment]
I would think it might be worth considering adding core backup/restore support for local plugins. Although, it might be difficult since local plugins exist outside of a course context.

**Comment by nadavkav** [comment]
Until we have core support, maybe... we can have a dummy metadata module that we add to each course with only backup and restore implemented?

**Comment by nadavkav** [comment]
Interesting: https://tracker.moodle.org/browse/MDL-55943

**Comment by nadavkav** [comment]
Found a promissing example for implementation :-)
https://github.com/synergylearning/moodle-local_filterurlresbak/tree/master/backup/moodle2

**Comment by mchurchward** [comment]
Since core supports it, we can add it. I would suggest the correct places to implement it would be in each context subplugin.

**Comment by nadavkav** [comment]
Yep. like mod/assign subplugins did.

**Comment by mchurchward** [comment]
@nadavkav are you going to implement any of the backup/restore code? Or are you looking for me to do that? Just setting my plans.

**Comment by nadavkav** [comment]
We are (one of my team, or me). sorry for no being clear on this.

**Comment by kirildevlion** [comment]
@mchurchward Hi, i tried to implement backup/restore, with local plugins. It seems that moodle does not add backup nested elements to XML in local metadata plugin. Maybe you can advice how to implement it? Thanks!
My commit here:
https://github.com/devlionco/moodle-local_metadata/commit/e0910166db25f39631cfd7aa614e18272d6592c2

**Comment by nadavkav** [comment]
@kirildevlion , please give a more descriptive name to process_localkiril :smile:

**Comment by seb-vial** [comment]
Hey,

I've managed to implement backup/restore for the plugin. BUT

- Course metadata is saved and restored
- Module metadata is saved and restored

And that's it...
What I can't figure out is how to backup/restore groups metadata.

For other metadata types (users, categories, cohorts), I don't think that's possible and more to the point, I think that's not useful as they can be used in courses, but their metadata is not specific to each course.
Also I've tried to implement backup/restore in each subplugin as suggested, but what I did had no effect. From what I understood from the documentation and Moodle code, the main plugin has to create its own xml file, and then subplugins can hook into that file. IMHO metadata should not be in a different file but injected where it belongs.
So I simply created 2 files :
- `backup_local_metadata_plugin.class.php` located in `/moodle-site/local/metadata/backup/moodle2` to handle backup for course metadata and module metadata. Each metadata is added in the according xml file (`course/course.xml` for course metadata and `activities/activity_xxx/module.xml`) in the generated zip file. I just haven't been able to backup group metadata in `groups.xml`
- `restore_local_metadata_plugin.class.php` located in `/moodle-site/local/metadata/backup/moodle2` which does the inverse, it's called when you restore a course and restore course metadata and module metadata.

**Comment by nadavkav** [comment]
Thank you @Birssan for sharing this code! I have downloaded it and it work great in our Moodle instance.
(When backing up a course and a module, of course. we are not using MD on other context levels)

**Comment by seb-vial** [comment]
It works fine for our Moodle as well, I guess it needs further testing though.

We only use metadata for modules as you do, but It would be nice if this feature is available for every metadata types.

**Comment by mchurchward** [comment]
Finally getting around to looking at this. @Birssan where is the code you created?

**Comment by seb-vial** [comment]
@mchurchward I haven't made a PR yet, I actually wanted to get your input on what I explained first. How come it's only possible to backup/restore course and module metadata but not the rest ?
I just pushed it to my fork if you want to check it out https://github.com/Birssan/moodle-local_metadata/commit/d4a2312c0e2f0e486f8979a7fe4548bcdde12a31
I now wonder how and where did @nadavkav get *my code* for this ?

**Comment by nadavkav** [comment]
@Birssan I got it from your repo, as far as I remember.
We use it. and it works fine.
We also have a fileupload field type, that gets a backup/restore for the file on the module context level.
It's on a private gitlab, in our campus, so I will push it into a public repo, so you can all enjoy it too. (hopefully, this weekend)

**Comment by seb-vial** [comment]
That's what I thought too, but turns out I hadn't pushed it, so I only pushed it yesterday. Anyway, not really important, you have it and it works, that's enough :p

**Comment by mchurchward** [comment]
Sorry I'm so late to the party here. This would still be a great addition.
Some questions:

1. If you restore a metadata item that does not exist in the system when it is restored, what happens?

1. How does the restore know that the data it is restoring is associated with the correct metadata item? Does it compare shortnames and fieldtypes? I wonder how custom user and course fields handle this...?

**Comment by susyzan** [comment]
Hi! Is anybody able to give an update about this feature please?
